storage/innobase/lock/lock0lock.cc

```diff
@@ -4692,61 +4820,92 @@ void lock_release_on_prepare(trx_t *trx)
-      if (!lock->is_table())
+      ut_ad(!lock->index->table->is_temporary());
+      if (!lock->is_rec_granted_exclusive_not_gap())
+        lock_rec_dequeue_from_page(lock, false);
+      else if (UNIV_UNLIKELY(lock->type_mode &
+                             (LOCK_PREDICATE | LOCK_PRDT_PAGE)))
+        /* SPATIAL INDEX locking is broken. */;
+      else
       {
-        ut_ad(!lock->index->table->is_temporary());
-        if (!lock->is_rec_granted_exclusive_not_gap())
-          lock_rec_dequeue_from_page(lock, false);
-        else if (UNIV_UNLIKELY(lock->type_mode &
-                               (LOCK_PREDICATE | LOCK_PRDT_PAGE)))
-          /* SPATIAL INDEX locking is broken. */;
-        else
+        auto cell= lock_sys.rec_hash.cell_get(lock->un_member.rec_lock.
+                                              page_id.fold());
+        if (lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM))
+          lock_rec_unlock(*cell, lock, PAGE_HEAP_NO_SUPREMUM);
+        else if (unlock_unmodified)
         {
-          auto cell= lock_sys.rec_hash.cell_get(lock->un_member.rec_lock.
-                                                page_id.fold());
-          if (lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM))
-            lock_rec_unlock(*cell, lock, PAGE_HEAP_NO_SUPREMUM);
-          else
+          ut_ad(lock->trx->isolation_level > TRX_ISO_READ_COMMITTED ||
+                /* Insert-intention lock is valid for supremum for isolation
+                   level > TRX_ISO_READ_COMMITTED */
+                lock->mode() == LOCK_X ||
+                !lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM));
+          ulint set_nth_bit_calls= trx->lock.set_nth_bit_calls;
+          lock_sys.wr_unlock();
+          trx->mutex_unlock();
+          mtr.start();
+          buf_block_t *block= btr_block_get(
+              *lock->index, lock->un_member.rec_lock.page_id.page_no(),
+              RW_S_LATCH, true, &mtr, nullptr, nullptr
+#if defined(UNIV_DEBUG) || !defined(DBUG_OFF)
+              , BUF_GET_POSSIBLY_FREED
+#endif /* defined(UNIV_DEBUG) || !defined(DBUG_OFF) */
+          );
+          lock_sys.wr_lock(SRW_LOCK_CALL);
+          if (block)
           {
-            ut_ad(lock->trx->isolation_level > TRX_ISO_READ_COMMITTED ||
-                  /* Insert-intention lock is valid for supremum for isolation
-                     level > TRX_ISO_READ_COMMITTED */
-                  lock->mode() == LOCK_X ||
-                  !lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM));
-            lock_rec_unlock_unmodified(*cell, lock, offsets, heap, mtr);
+            /* concurent lock_sys_t::resize() call could make a previously
+            computed cell address invalid */
+            cell= lock_sys.rec_hash.cell_get(lock->un_member.rec_lock.
+                                             page_id.fold());
+            (void) lock_rec_unlock_unmodified(block, cell, lock, offsets, heap,
+                                              GLOBAL);
```

Comment:
@mariadb-DebarunBanerjee recently taught us that the correct C++11 way to declare that a result is being discarded is to assign it to `std::ignore`:
```c++
            std::ignore=
              lock_rec_unlock_unmodified(block, cell, lock, offsets_, GLOBAL);
```
See my other comment regarding removing `offsets` and `heap` from the caller.